### DISCUSSION TITLE
http keep alive & tcp keep alive

### DISCUSSION THREAD

**Discussion by yeonkkk** [woowacourse-study/2023-tal-study#66] [기술 면접 질답 모음집]
http keep alive & tcp keep alive  에 대해 설명해주세요

**Comment by yeonkkk** [comment]
HTTP Keep-Alive는 웹 브라우저와 서버 간의 연결을 재사용하여 성능 향상을 위한 것이며,
TCP Keep-Alive는 TCP 연결의 유효성을 검사하여 연결 리소스를 관리하는 역할입니다. 

둘의 설정을 살펴보면 HTTP Keep-alive의 경우는 ‘기능의 활성화 여부’ 그리고 ‘Timeout 시간’만을 설정할 수 있고, 
TCP Keep-alive의 경우는 ‘Keep-alive 유지 시간’, ‘Keep-alive 패킷 재전송 시간(interval)’, ‘패킷 전송 빈도수’를 설정할 수 있습니다.

즉, HTTP Keep-alive의 경우에는 ‘일정 시간 기다려본다.’라는 의미가 강하고, TCP Keep-alive 는  ‘연결을 유지한다.’ 라는 의미가 강합니다.

HTTP Keep-Alive는 HTTP/1.1부터 기본으로 활성화되며 Connection 헤더를 통해 조절할 수 있습니다.
TCP Keep-Alive는 운영 체제나 TCP 스택에서 구성 가능하며 기본적으로 활성화되어 있으며 연결이 여전히 활성 상태인지 확인합니다.

**Reply by yeonkkk** [reply]
ref. https://dabeen.medium.com/becoming-se-what-is-keep-alive-478b1ea99593

**Comment by yoondgu** [comment]
HTTP Keep-Alive는 HTTP 요청 - 응답 간 헤더 정보를 활용하여 일정 시간동안 연결을 유지시키는 것입니다.
이를 통해 매 요청마다 새 TCP 연결을 만드는 대신, 기존 연결을 닫지 않고 재사용할 수 있습니다.  
HTTP 1.1은 이를 기본적으로 지원하지만, HTTP 1.0은 직접 Connection 헤더에 keep-alive 옵션을 명시해주어야 합니다.
timeout 값을 통해 얼마 동안 해당 연결을 유지시킬지 정할 수 있습니다.  

TCP Keep-Alive는 두 종단 간의 연결이 유효한지 확인하고 관리하는 것입니다.
일정 시간이 지나면 3-way handshake로 연결이 살아 있는지(클라이언트가 응답 가능한지) 확인하고, 응답이 없으면 연결을 끊는 방식으로 관리합니다.

ref. https://devidea.tistory.com/60
